rasa/core/actions/flows.py

```diff
@@ -73,3 +80,84 @@ async def run(
             events.append(ActiveLoop(None))
 
         return events
+
+
+class ActionCancelFlow(action.Action):
+    """Action which cancels a flow from the stack."""
+
+    def __init__(self) -> None:
+        """Creates a `ActionCancelFlow`."""
+        super().__init__()
+
+    def name(self) -> Text:
+        """Return the flow name."""
+        return ACTION_CANCEL_FLOW
+
+    async def run(
+        self,
+        output_channel: "OutputChannel",
+        nlg: "NaturalLanguageGenerator",
+        tracker: "DialogueStateTracker",
+        domain: "Domain",
+        metadata: Optional[Dict[Text, Any]] = None,
+    ) -> List[Event]:
+        """Cancel the flow."""
+        stack = FlowStack.from_tracker(tracker)
+        if stack.is_empty():
+            structlogger.warning("action.cancel_flow.no_active_flow", stack=stack)
+            return []
+
+        context = tracker.get_slot(FLOW_CONTEXT_SLOT) or {}
+        canceled_flow_frames = context.get("canceled_frames", [])
+
+        for frame_idx in canceled_flow_frames:
+            if frame_idx >= len(stack.frames):
+                structlogger.warning(
+                    "action.cancel_flow.frame_not_found",
+                    stack=stack,
+                    frame_idx=frame_idx,
+                )
+                continue
+            stack.frames[frame_idx].step_id = END_STEP
```

Comment:
I don't like that we are referencing the frames using only their index here. I think there's potential for stuff to happen in between the cancel frame was put on the stack and this cancel action being executed ~ which could jumble everything.. Could stack frames get a more resilient ID?

Reply:
Also, the warning here doesn't seem to help to resolve the issue. It's indicating that the stack frame has changed, and still we are continuing with the other indices to manipulate stack frames 😬

Reply:
🙏 yeah that should do the trick

Reply:
I've changed the implementation and added ids to every stackframe.